[Sema] Consolidate logic for whether a decl has a curried self or parameter list

hamishknight opened the pull request:
This PR adds `ValueDecl::hasCurriedSelf` and `ValueDecl::hasParameterList` that return whether a given declaration has a curried self parameter in its interface type and whether it has a parameter list, respectively. These members are then used in a couple of places where this logic is duplicated, as well as to fix a code completion bug.

In addition, this PR tweaks `isDeclAsSpecializedAs` so that it could compare decls that differ by curry level if it needed to. This case doesn't currently come up though, so this should be NFC.

Finally, this PR has a couple of straightforward NFC cleanups.

Resolves [SR-10795](https://bugs.swift.org/browse/SR-10795).

hamishknight committed:
[Sema][NFC] Remove dead function

hamishknight committed:
[Sema][NFC] Simplify paramIsIUO implementation

hamishknight committed:
[AST][Sema] Add ValueDecl::hasParameterList

hamishknight committed:
[AST][Sema] Add ValueDecl::hasCurriedSelf

Use this function to replace various places where the logic is
duplicated.

In addition, isolate the logic where subscripts are treated as having
curried self parameters to CalleeCandidateInfo, as their interface types
don't have a curried self, but get curried with self by
CalleeCandidateInfo. Ideally we'd fix this by having a subscript's
interface type be curried with self, but given that most of this CSDiag
logic should be going away, this may not be necessary.

hamishknight committed:
[CodeCompletion] Fix a call to computeDefaultMap

Resolves SR-10795.

hamishknight committed:
[CSRanking] Handle mismatching curry levels

While it's currently not possible for `isDeclAsSpecializedAs` to compare
decls that differ by having a curried self or having a parameter list,
tweak the logic slightly so it could handle that case if it needed to.

If one decl has a parameter list and the other doesn't, then just
compare their self types. If one decl is curried and the other isn't,
then add a curried self to the other and compare as normal.

slavapestov commented:
Cool!

Is it worth using this in `SILDeclRef::getParameterListCount()`?

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -855,34 +848,23 @@ getCalleeDeclAndArgs(ConstraintSystem &cs,
+    // parameter, it gets dropped from the type of the reference.
+    bool hasAppliedSelf =
+        decl->hasCurriedSelf() && doesDeclRefApplyCurriedSelf(baseType, decl);
+    return std::make_tuple(decl, hasAppliedSelf, argLabels, hasTrailingClosure);
```
I wonder if it would be cleaner to return `choice` form here and add `baseType` parameter to `computeDefaultMap` so it could take care of figuring out whether self has been applied or not instead of relying on caller to do so?

hamishknight replied:
@xedin Thanks for the review! Unfortunately, I don't believe this will be easy for `coerceCallArguments`, which calls `computeDefaultMap` with slightly different logic for `skipCurriedSelf` – for example, passing `false` for a `SelfApplyExpr`.

xedin replied:
Fair enough.

xedin reviewed (approved):
LGTM, thank you! I have left a small comment inline.

hamishknight committed:
[AST] Add ValueDecl::getNumCurryLevels

hamishknight commented:
@slavapestov Good idea! I actually just noticed that we're duplicating the logic for how many curry levels a decl has in a bunch of places, so added `ValueDecl::getNumCurryLevels`.

slavapestov commented:
@hamishknight Thanks!

slavapestov reviewed (approved):
This looks great

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

hamishknight commented:
LLDB failure for OS X looks unrelated, also observed in https://ci.swift.org/job/swift-PR-osx/13822/.

slavapestov commented:
@swift-ci Please smoke test macOS

hamishknight committed:
[Sema] Don't assume member refs have base types

- Add a precondition on `doesDeclRefApplyCurriedSelf` to expect
a member decl, and rename it to make the precondition explicit.

- Don't assume that not having a base type means this isn't a member
reference, as member references to static operators don't have base
types.

Resolves SR-10843.

hamishknight commented:
Just found a 5.1 regression ([SR-10843](https://bugs.swift.org/browse/SR-10843)) related to this logic, so pushed a fix (I'll also open a 5.1 PR in a bit). Could you please re-run the CI when it comes back up?

hamishknight commented:
Opened a 5.1 PR for SR-10843: https://github.com/apple/swift/pull/25288.

hamishknight committed:
Merge remote-tracking branch 'upstream/master' into a-couple-of-tangents

xedin commented:
@swift-ci please test

hamishknight commented:
Linux failure appears to be unrelated

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge

hamishknight committed:
Merge remote-tracking branch 'upstream/master' into a-couple-of-tangents

hamishknight commented:
Had to resolve a merge conflict with https://github.com/apple/swift/pull/25359, could you please try again?

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
